File: Quicksort/go/quicksort.go
package main

import "fmt"

func partition(elements []int, start int, end int) int {
	marker := start
	right := end
	preMarker := start - 1
	for {
		if marker == end {
			break
		}
		if elements[marker] <= elements[right] {
			preMarker++
			elements[preMarker], elements[marker] = elements[marker], elements[preMarker]
		}
		marker++
	}
	elements[preMarker+1], elements[right] = elements[right], elements[preMarker+1]
	return preMarker + 1
}

func quickSort(elements []int, start int, end int) {
	if start < end {
		loc := partition(elements, start, end)
		quickSort(elements, start, loc-1)

		quickSort(elements, loc+1, end)

	}
}

func main() {
	elements := make([]int, 8)
	elements = []int{2, 8, 7, 1, 3, 5, 6, 4}
	quickSort(elements, 0, 7)
	fmt.Println(elements)
}

File: Quicksort/python/quicksort.py

counter = 0


def partition(elements, start, end) -> int:
    global counter
    preMarker = start-1
    marker = start

    while marker < end:
        counter += 1
        if elements[end] >= elements[marker]:
            preMarker += 1
            elements[marker], elements[preMarker] = elements[preMarker], elements[marker]
        marker += 1
    elements[end], elements[preMarker+1] = elements[preMarker+1], elements[end]
    return preMarker+1


def quickSort(elements, start, end):
    if start < end:
        loc = partition(elements, start, end)
        quickSort(elements, start, loc-1)
        quickSort(elements, loc+1, end)


if __name__ == "__main__":
    numbers = [11,21,9,10,5]
    print(numbers)
    quickSort(numbers, 0, len(numbers)-1)
    print(numbers, counter, len(numbers))
